scripts/bluetooth/mesh/mesh_dfu_metadata.py

```diff
@@ -0,0 +1,388 @@
+            "sig_models": self.sig_list,
+            "vendor_models": self.vnd_list,
+        }
+
+
+class Comp0:
+    # Must stay in order
+    FEATURE_KCONF_OPTS = [
+        'CONFIG_BT_MESH_RELAY',
+        'CONFIG_BT_MESH_GATT_PROXY',
+        'CONFIG_BT_MESH_FRIEND',
+        'CONFIG_BT_MESH_LOW_POWER',
+    ]
+
+    def __init__(self, cid, pid, vid, kconfig):
+        self.elems = []
+        self.cid = cid
+        self.pid = pid
+        self.vid = vid
+        self.__features_add(kconfig)
+        self.hash = None
+
+    def elem_add(self, loc):
+        new_elem = Elem(loc)
+        self.elems.append(new_elem)
+        return new_elem
+
+    def __features_add(self, kconfig):
+        if 'CONFIG_BT_MESH_CRPL' not in kconfig.keys():
+            exit_with_error_msg("Could not find CONFIG_BT_MESH_CRPL Kconfig option")
+
+        self.feat = 0
+        self.crpl = int(kconfig['CONFIG_BT_MESH_CRPL'])
+
+        for i, opt in enumerate(self.FEATURE_KCONF_OPTS):
+            self.feat += (1 if kconfig.get(opt) == 'y' else 0) << i
+
+    def __bytestring_generate(self):
+        bytestring = bytearray()
+        bytestring.extend(self.cid.to_bytes(2, 'little'))
+        bytestring.extend(self.pid.to_bytes(2, 'little'))
+        bytestring.extend(self.vid.to_bytes(2, 'little'))
+        bytestring.extend(self.crpl.to_bytes(2, 'little'))
+        bytestring.extend(self.feat.to_bytes(2, 'little'))
+
+        for elem in self.elems:
+            bytestring.extend(elem.bytestring_generate())
+
+        return bytestring
+
+    def dict_generate(self):
+        return {
+            "cid": self.cid,
+            "pid": self.pid,
+            "vid": self.vid,
+            "crpl": self.crpl,
+            "features": self.feat,
+            "elements": [e.dict_generate() for e in self.elems]
+        }
+
+    def hash_generate(self):
+        # Uses 16-byte zero key
+        crypto = cmac.CMAC(algorithms.AES(bytes(16)))
+        crypto.update(bytes(self.__bytestring_generate()))
+        self.hash, *_ = struct.unpack('<L', crypto.finalize()[:4])
+        return self.hash
+
+
+class KConfig(dict):
+
+    @classmethod
+    def from_file(cls, filename):
+        """
+        Extracts content of '.config' file into a dictionary
+
+        Parameters:
+            config_path (str): Path to '.config' file used by a firmware
+        Returns:
+            cls: A KConfig object where keys are Kconfig option names and values
+            are option values, parsed from the config file at config_path.
+        """
+        configs = cls()
+        try:
+            with open(filename, 'r') as config:
+                for line in config:
+                    if not line.startswith("CONFIG_"):
+                        continue
+                    kconfig, value = line.split("=")
+                    configs[kconfig] = value.strip()
+                return configs
+        except Exception:
+            exit_with_error_msg("Unable to parse .config file")
```

Comment:
Maybe worth printing exception message in all try-except cases to know exactly what happened?

Reply:
Im trying something like this:
```
except Exception as err:
exit_with_error_msg(f"Failed to extract composition data from .elf file: \n\t{err}")
```
That will add the subreason for the fail as well.

Reply:
Sounds good! Please try it once (manually raise exception at any place) just to see how it prints the message and the exception.

Reply:
I agree with Pavel that exceptions should be raised up to main. Then, for exceptions where can't add any more info than the original exception (which IMO are most of them), remove the `try ... except` and let main handle the original exception. If you really need to report more info, do 
```python
try:
    do_thing_that_can_fail()
except Exception as err:
    raise Exception("Thing that can fail has failed") from err
```

Reply:
When doing this, `f"{err}"` returns `"Thing that can fail has failed"`. To get the string for the sub-exception, do `f"{err.__cause__}"` (after checking for `err.__cause__ is None`)

Reply:
Or, easier, `import traceback` and `print_exception`/`print_exc` and let Python do the work of printing the relevant messages, line numbers etc.

Reply:
@Andrewpini you seem to have misunderstood me. What I meant was: `exit_with_error_msg` should be called from exactly one place, an `except` clause in main. All other errors should just be left to propagate, or chained using `raise Exception() from err`.

Meaning, for this piece of code specifically (but also every other place currently calling `exit_with_error_msg` from deep within some function), you'd do

```python
try:
    with open(filename, 'r') as config:
        for line in config:
            if not line.startswith("CONFIG_"):
                continue
            kconfig, value = line.split("=")
            configs[kconfig] = value.strip()
        return configs
except Exception as err:
    raise Exception("Unable to parse .config file") from err
```

and then main would be

```python
try:
    kconfigs = KConfig.from_file(config_path)
    comp = parse_comp_data(elf_path, kconfigs)
    version = kconfigs.version_parse()
    encoded_metadata = encoded_metadata_get(version, comp)
except Exception as err:
    # Do something here, either traceback.print_exc + sys.exit(0), print(err) + sys.exit etc.
```

but really, unless you really care about returning 0 in sys.exit, just letting any exception propagate out of main will accomplish the same thing: The error is printed and the script exits (but with exit code 1). This is currently what will happen in case something goes wrong with the zip-file operations, for instance.

Reply:
@PavelVPV @Andrewpini resolving this since I think it is handled fully now